src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -162,29 +163,64 @@ private LanguageExpression GetResourceNameExpression(ResourceDeclarationSyntax r
                 // this condition should have already been validated by the type checker
                 throw new ArgumentException($"Expected resource syntax to have type {typeof(ObjectSyntax)}, but found {resourceSyntax.Body.GetType()}");
             }
+            
+            // this condition should have already been validated by the type checker
+            return TryGetNamedPropertyExpression(objectSyntax, "name") ?? throw new ArgumentException($"Expected resource syntax body to contain property 'name'");
+        }
 
-            var namePropertySyntax = objectSyntax.Properties.FirstOrDefault(p => LanguageConstants.IdentifierComparer.Equals(p.TryGetKeyText(), "name"));
-            if (namePropertySyntax == null)
+        private LanguageExpression GetModuleNameExpression(ModuleDeclarationSyntax moduleSyntax)
+        {
+            if (!(moduleSyntax.Body is ObjectSyntax objectSyntax))
             {
                 // this condition should have already been validated by the type checker
-                throw new ArgumentException($"Expected resource syntax body to contain property 'name'");
+                throw new ArgumentException($"Expected module syntax to have type {typeof(ObjectSyntax)}, but found {moduleSyntax.Body.GetType()}");
+            }
+            
+            // this condition should have already been validated by the type checker
+            return TryGetNamedPropertyExpression(objectSyntax, "name") ?? throw new ArgumentException($"Expected module syntax body to contain property 'name'");
+        }
+
+        /// <summary>
+        /// Tries to get an ARM expression for a named property from a source object syntax. Returns null if no property is found.
+        /// </summary>
+        private LanguageExpression? TryGetNamedPropertyExpression(ObjectSyntax objectSyntax, string propertyName)
+        {
+            var namePropertySyntax = objectSyntax.Properties.FirstOrDefault(p => LanguageConstants.IdentifierComparer.Equals(p.TryGetKeyText(), propertyName));
+            if (namePropertySyntax == null)
+            {
+                return null;
             }
 
             return ConvertExpression(namePropertySyntax.Value);
         }
 
+        private static FunctionExpression GenerateResourceIdExpression(SemanticModel.SemanticModel semanticModel, string fullyQualifiedType, IEnumerable<LanguageExpression> nameSegments)
+        {
+            var resourceIdFunctionName = semanticModel.TargetScope switch {
+                ResourceScopeType.TenantScope => "tenantResourceId",
+                ResourceScopeType.ManagementGroupScope => "managementGroupResourceId", // TODO - this doesn't really exist!
```

Comment:
TODO - what should we codegen here?

Reply:
Can we generate a call to the tenantResourceId function or just construct something via concat?

Reply:
I'm imagining it would be something like, but trying to get confirmation from Brian on the idiomatic way to do it:
`[extensionResourceId(deployment().scope, 'Microsoft.Resources/deployments', deploymentName)]`

Reply:
ok, summary of the conversation - we should be able to codegen:

```
"dependsOn": [
  "[concat('{value of the scope property}', 'Microsoft.Resources/deployments/', '{deploymentName}')]
]
```

We're pretty sure that dependsOn doesn't need a fully qualified resourceId, but a relative one.  I think that works for all scenarios...